File: frontend/src/services/curriculumService.ts
import api from '../lib/api';

export interface Curriculum {
  id: number;
  title: string;
  description: string;
  grade_level: string;
  educational_level_id: number;
  subject_id: number;
  curriculum_standard: string;
  academic_year: string;
  term: string;
  created_by: number;
  status: string;
  duration_weeks?: number;
  created_at: string;
  updated_at: string;
  subject_name?: string;
}

export interface CurriculumUnit {
  id: number;
  curriculum_id: number;
  title: string;
  description: string;
  objectives: string;
  resources: string;
  duration_weeks: number;
  sequence_order: number;
  created_at: string;
  updated_at: string;
}

export interface CurriculumCreate {
  title: string;
  description: string;
  educational_level_id: number;
  subject_id: number;
  curriculum_standard: string;
  academic_year: string;
  term: string;
  duration_weeks?: number;
  status?: string;
}

export interface CurriculumUnitCreate {
  curriculum_id: number;
  title: string;
  description: string;
  objectives: string;
  resources: string;
  duration_weeks: number;
  sequence_order: number;
}

const curriculumService = {
  getCurricula: async (params?: {
    subject_id?: number;
    educational_level_id?: number;
    academic_year?: string;
    term?: string;
  }): Promise<Curriculum[]> => {
    try {
      const response = await api.get('/curriculum', { params });
      return ((response.data?.curricula || response.data?.data || []) as any[]).map((curriculum) => ({
        ...curriculum,
        grade_level: curriculum.grade_level || curriculum.educational_level?.name || '',
        subject_name: curriculum.subject_name || curriculum.subject?.name,
      }));
    } catch (error) {
      console.error('Error fetching curricula:', error);
      throw error;
    }
  },

  getCurriculumById: async (curriculumId: number): Promise<Curriculum> => {
    try {
      const response = await api.get(`/curriculum/${curriculumId}`);
      const curriculum = response.data?.curriculum || response.data?.data;
      return {
        ...curriculum,
        grade_level: curriculum?.grade_level || curriculum?.educational_level?.name || '',
        subject_name: curriculum?.subject_name || curriculum?.subject?.name,
      };
    } catch (error) {
      console.error(`Error fetching curriculum ${curriculumId}:`, error);
      throw error;
    }
  },

  createCurriculum: async (curriculumData: CurriculumCreate): Promise<Curriculum> => {
    try {
      const response = await api.post('/curriculum/', curriculumData);
      const curriculum = response.data?.curriculum || response.data?.data;
      return {
        ...curriculum,
        grade_level: curriculum?.grade_level || curriculum?.educational_level?.name || '',
        subject_name: curriculum?.subject_name || curriculum?.subject?.name,
      };
    } catch (error) {
      console.error('Error creating curriculum:', error);
      throw error;
    }
  },

  getCurriculumUnits: async (curriculumId: number): Promise<CurriculumUnit[]> => {
    try {
      const response = await api.get(`/curriculum/${curriculumId}/units`);
      return response.data?.units || response.data?.data || [];
    } catch (error) {
      console.error(`Error fetching units for curriculum ${curriculumId}:`, error);
      throw error;
    }
  },

  createCurriculumUnit: async (unitData: CurriculumUnitCreate): Promise<CurriculumUnit> => {
    try {
      const response = await api.post(`/curriculum/${unitData.curriculum_id}/units`, unitData);
      return response.data?.unit || response.data?.data;
    } catch (error) {
      console.error('Error creating curriculum unit:', error);
      throw error;
    }
  },

  updateCurriculum: async (curriculumId: number, curriculumData: Partial<CurriculumCreate>): Promise<Curriculum> => {
    try {
      const response = await api.put(`/curriculum/${curriculumId}`, curriculumData);
      const curriculum = response.data?.curriculum || response.data?.data;
      return {
        ...curriculum,
        grade_level: curriculum?.grade_level || curriculum?.educational_level?.name || '',
        subject_name: curriculum?.subject_name || curriculum?.subject?.name,
      };
    } catch (error) {
      console.error(`Error updating curriculum ${curriculumId}:`, error);
      throw error;
    }
  },

  deleteCurriculum: async (curriculumId: number): Promise<void> => {
    try {
      await api.delete(`/curriculum/${curriculumId}`);
    } catch (error) {
      console.error(`Error deleting curriculum ${curriculumId}:`, error);
      throw error;
    }
  },

  updateCurriculumUnit: async (unitId: number, unitData: Partial<CurriculumUnitCreate>): Promise<CurriculumUnit> => {
    try {
      const response = await api.put(`/curriculum/units/${unitId}`, unitData);
      return response.data?.unit || response.data?.data;
    } catch (error) {
      console.error(`Error updating curriculum unit ${unitId}:`, error);
      throw error;
    }
  },

  deleteCurriculumUnit: async (unitId: number): Promise<void> => {
    try {
      await api.delete(`/curriculum/units/${unitId}`);
    } catch (error) {
      console.error(`Error deleting curriculum unit ${unitId}:`, error);
      throw error;
    }
  }
};

export default curriculumService;

File: backend/app/services/curriculum_service.py
# Create a new service for curriculum operations
from app.models.curriculum import Curriculum
from app.models.curriculum_unit import CurriculumUnit
from app.extensions import db
from sqlalchemy import desc

class CurriculumService:
    """Service for curriculum-related operations."""
    
    @staticmethod
    def create_curriculum(data):
        """Create a new curriculum."""
        try:
            curriculum = Curriculum(**data)
            db.session.add(curriculum)
            db.session.commit()
            return curriculum, None
        except Exception as e:
            db.session.rollback()
            return None, str(e)
    
    @staticmethod
    def add_curriculum_unit(data):
        """Add a unit to a curriculum."""
        try:
            unit = CurriculumUnit(**data)
            db.session.add(unit)
            db.session.commit()
            return unit, None
        except Exception as e:
            db.session.rollback()
            return None, str(e)
    
    @staticmethod
    def get_curriculum_by_subject(subject_id, grade_level=None, academic_year=None):
        """Get curricula for a specific subject."""
        query = Curriculum.query.filter(Curriculum.subject_id == subject_id)
        
        if grade_level:
            query = query.filter(Curriculum.grade_level == grade_level)
            
        if academic_year:
            query = query.filter(Curriculum.academic_year == academic_year)
            
        return query.order_by(Curriculum.title).all()
    
    @staticmethod
    def get_all_curricula(page=1, per_page=20, subject_id=None, grade_level=None, academic_year=None):
        """Get all curricula with optional filtering."""
        query = Curriculum.query
        
        if subject_id:
            query = query.filter(Curriculum.subject_id == subject_id)
            
        if grade_level:
            query = query.filter(Curriculum.grade_level == grade_level)
            
        if academic_year:
            query = query.filter(Curriculum.academic_year == academic_year)
            
        return query.order_by(desc(Curriculum.created_at)).paginate(page=page, per_page=per_page)
    
    @staticmethod
    def get_curriculum_by_id(curriculum_id):
        """Get a curriculum by ID."""
        return Curriculum.query.get(curriculum_id)
    
    @staticmethod
    def update_curriculum(curriculum_id, data):
        """Update a curriculum."""
        try:
            curriculum = Curriculum.query.get(curriculum_id)
            if not curriculum:
                return None, "Curriculum not found"
                
            for key, value in data.items():
                setattr(curriculum, key, value)
                
            db.session.commit()
            return curriculum, None
        except Exception as e:
            db.session.rollback()
            return None, str(e)
    
    @staticmethod
    def delete_curriculum(curriculum_id):
        """Delete a curriculum."""
        try:
            curriculum = Curriculum.query.get(curriculum_id)
            if not curriculum:
                return False, "Curriculum not found"
                
            db.session.delete(curriculum)
            db.session.commit()
            return True, None
        except Exception as e:
            db.session.rollback()
            return False, str(e)
    
    @staticmethod
    def get_curriculum_units(curriculum_id):
        """Get all units for a curriculum."""
        return CurriculumUnit.query.filter_by(curriculum_id=curriculum_id).order_by(CurriculumUnit.sequence_order).all()
    
    @staticmethod
    def get_curriculum_unit_by_id(unit_id):
        """Get a curriculum unit by ID."""
        return CurriculumUnit.query.get(unit_id)
    
    @staticmethod
    def update_curriculum_unit(unit_id, data):
        """Update a curriculum unit."""
        try:
            unit = CurriculumUnit.query.get(unit_id)
            if not unit:
                return None, "Curriculum unit not found"
                
            for key, value in data.items():
                setattr(unit, key, value)
                
            db.session.commit()
            return unit, None
        except Exception as e:
            db.session.rollback()
            return None, str(e)
    
    @staticmethod
    def delete_curriculum_unit(unit_id):
        """Delete a curriculum unit."""
        try:
            unit = CurriculumUnit.query.get(unit_id)
            if not unit:
                return False, "Curriculum unit not found"
                
            db.session.delete(unit)
            db.session.commit()
            return True, None
        except Exception as e:
            db.session.rollback()
            return False, str(e)